### ISSUE TITLE
Dropout2d uses a lot of memory.

### ISSUE THREAD

**Issue by jatentaki** [pytorch/pytorch#16577]
## 🐛 Bug
`Dropout2d` uses a lot of memory. For a `B, F, H, W` tensor, it could get away with `B * F` bits to store whether each feature/batch item was kept or not. Instead, it appears to store something proportional to `B * F * H * W`. According to @apaszke this is because it is implemented with multiplication rather than `masked_fill` or `where` (since this makes it easier to switch between `eval` and `train`). A solution would be to have different control paths for `train` and `eval` or
> an easy fix would be to improve the derivative code we generate for multiplies such that they don’t save `x` when you do `x * y` with `y` that doesn’t need gradient

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 396.54
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.6            py36hd81dba3_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.0.0           py3.6_cuda9.0.176_cudnn7.4.1_1    pytorch
[conda] tensorflow                1.12.0          mkl_py36h69b6ba0_0  
[conda] tensorflow-base           1.12.0          mkl_py36h3c3e929_0  
[conda] torchvision               0.2.1                      py_2    pytorch
